### PR TITLE
Add testing strategy and developer guideline

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area testing dev-productivity
/kind enhancement

**What this PR does / why we need it**:

Manifest our testing strategy and a guideline for developing tests.
This summarizes and consolidates the results from two workshops on the topic with @gardener/gardener-maintainers.
Thanks again everybody for participating, collecting thoughts, and discussing this topic!

**Which issue(s) this PR fixes**:
Part of https://github.com/gardener/gardener/issues/6016

**Special notes for your reviewer**:

**Release note**:

```feature developer
A new testing strategy and developer guideline has been added. Make sure to check out the [document](https://github.com/gardener/gardener/blob/master/docs/development/testing.md#writing-test-machinery-tests) if you want to learn more about the different kinds of tests we use and how to best write them!
```

### CODE REVIEW

File: docs/development/testing.md [hunk 1]
```diff
@@ -1,26 +1,194 @@
...
+- what kind of tests we have in Gardener
+- how to run each of them
+- what purpose each kind of test serves
+- how to best write tests that are correct, stable, fast and maintainable
+- how to debug tests that are not working as expected
+
+The document is aimed towards developers that want to contribute code and need to write tests, as well as maintainers and reviewers that review test code.
+It serves as a common guide that we commit to follow in our project to ensure consistency in our tests, good coverage for high confidence and good maintainability.
+
+The guidelines are not ment to be absolute rules.
```
Comment (rfranzke) [inline docs/development/testing.md]:
```suggestion
The guidelines are not meant to be absolute rules.
```

File: docs/development/testing.md [hunk 2]
```diff
@@ -1,26 +1,194 @@
...
+Always apply common sense and adapt the guideline if it doesn't make much sense for some cases.
+If in doubt, don't hesitate to ask questions during PR review (as author but also as reviewer).
+Add new learnings as soon as we make them!
+
+Generally speaking, **tests are a strict requirement for contributing new code**.
+If you touch code that is currently untested, you need to add tests for the new cases that you introduce as a minimum.
+Ideally though, you would add the missing test cases for the current code as well (**boy scout rule**).
```
Comment (rfranzke) [inline docs/development/testing.md]: Add a link to some description for the boy scout rule :)

File: docs/development/testing.md [hunk 3]
```diff
@@ -1,26 +1,194 @@
...
+- use the same client schemes that are also used by production code to avoid subtle bugs/regressions: [example PR](https://github.com/gardener/gardener/pull/5469/commits/b0d9e2baeaee246c6eb466a181df35adca4a6ada)
+- use the debugger to ensure your test code is actually asserting the right thing
+- about overwriting global variables
+  - this is a common pattern (or hack?) in go for faking calls to external functions
+  - however, this can lead to races, when the global variable is used from a goroutine (e.g., the function is called)
+  - alternatively, set fields on structs (passed via parameter or set directly): this is not racy, as struct values are typically (and should be) only used for a single test case
+  - alternative to dealing with function variables and fields:
+    - add an interface, which your code depends on
+    - write a fake and a real implementation (similar to `clock.Clock.Sleep`)
+    - the real implementation calls the actual function (`clock.RealClock.Sleep`, calls `time.Sleep`)
```
Comment (rfranzke) [inline docs/development/testing.md]:
```suggestion
    - the real implementation calls the actual function (`clock.RealClock.Sleep` calls `time.Sleep`)
```

File: docs/development/testing.md [hunk 4]
```diff
@@ -1,26 +1,194 @@
...
+SUCCESS! -- 3 Passed | 0 Failed | 0 Pending | 0 Skipped
+PASS
+
+All tests passed...
+Will keep running them until they fail.
+This was attempt #58
+No, seriously... you can probably stop now.
+```
+
+Use the [`stress` tool](https://pkg.go.dev/golang.org/x/tools/cmd/stress) for deflaking tests, that fail sporadically in CI, e.g., due resource contention (CPU throttling):
```
Comment (rfranzke) [inline docs/development/testing.md]:
```suggestion
Use the [`stress` tool](https://pkg.go.dev/golang.org/x/tools/cmd/stress) for deflaking tests that fail sporadically in CI, e.g., due resource contention (CPU throttling):
```

File: docs/development/testing.md [hunk 5]
```diff
@@ -1,26 +1,194 @@
...
+- test coverage can be used as a tool during test development for covering all cases of a unit
+- however, test coverage data can be a false safety net
+  - full line coverage doesn't mean you have covered all cases of valid input
+  - we don't have strict requirements for test coverage, as it doesn't necessarily yield the desired outcome
+- unit tests should not test too large components, e.g. entire controller `Reconcile` functions
+  - if a function/component does many steps, it's probably better to split it up into multiple functions/components that can be unit tested individually
+  - there might be special cases for very small `Reconcile` functions
+  - if there are a lot of edge cases, extract dedicated functions that cover them and use unit tests to test them
+  - usual-sized controllers should rather be tested in integration tests
+  - individual parts (e.g. helper functions) should still be tested in unit test though for covering all cases
```
Comment (rfranzke) [inline docs/development/testing.md]:
```suggestion
  - individual parts (e.g. helper functions) should still be tested in unit test for covering all cases, though
```

File: docs/development/testing.md [hunk 6]
```diff
@@ -1,26 +1,194 @@
...
+  - mocks decrease maintainability because they expect the tested component to follow a certain way to reach the desired goal (e.g., call specific functions with particular arguments), [example consequence](https://github.com/gardener/gardener/pull/4027/commits/111aba2c8e306421f2fa6b27e5d8ed8b2fc52be9#diff-8e61507edf985df2625840a690115c43bca6c032f2ff818389633bd4365c3efdR293-R298)
+  - generally, fakes should be used in "result-oriented" test code (e.g., that a certain object was labelled, but the test doesn't care if it was via patch or update as both a valid ways to reach the desired goal)
+  - although rare, there are valid use cases for mocks, e.g. if the following aspects are important for correctness:
+    - asserting that an exact function is called
+    - asserting that functions are called in a specific order
+    - asserting that exact parameters/values/... are passed
+    - asserting that a certain function was not called
+    - many of these can also be verified with fakes, although mocks might be simpler
+  - only use mocks, if the tested code directly calls the mock; never if the tested code only calls the mock indirectly (e.g., through a helper package/function)
```
Comment (rfranzke) [inline docs/development/testing.md]:
```suggestion
  - only use mocks if the tested code directly calls the mock; never if the tested code only calls the mock indirectly (e.g., through a helper package/function)
```

File: docs/development/testing.md [hunk 7]
```diff
@@ -49,9 +217,49 @@ export USE_EXISTING_CLUSTER=true
...
+- example component with multiple units: a controller with its reconciler, watches, predicates, event handlers, queues, etc.
+- integration tests set up a full component (including used libraries) and run it against a test environment close to the actual setup
+  - e.g., start controllers against a real Kubernetes control plane to catch bugs that can only happen when talking to a real API server
+  - integration tests are generally more expensive to run (e.g., in terms of execution time)
+- integration tests should not cover each and every detailed case
+  - rather cover a good portion of the "usual" cases that components will face during normal operation (positive and negative test cases)
+  - but don't cover all failure cases or all cases of predicates -> they should be covered in unit tests already
+  - generally, not supposed to "generate test coverage" but to provide confidence that components work well
+- as integration tests typically test only one component (or a cohesive set of components) isolated from others, they cannot catch bugs that occur, e.g., when multiple controllers interact (could be discovered by e2e tests though)
```
Comment (rfranzke) [inline docs/development/testing.md]:
```suggestion
- as integration tests typically test only one component (or a cohesive set of components) isolated from others, they cannot catch bugs that occur when multiple controllers interact (could be discovered by e2e tests though)
```

**Review by rfranzke — COMMENTED** [review]
/lgtm

File: docs/development/testing.md [hunk 8]
```diff
@@ -1,26 +1,194 @@
...
+- introduce helper functions for assertions to make test more readable where applicable: [example test](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/test/integration/gardenlet/shootsecret/controller_test.go#L323-L331)
+- introduce custom matchers to make tests more readable where applicable: [example matcher](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/pkg/utils/test/matchers/matchers.go#L51-L57)
+- don't rely on accurate timing of `time.Sleep` and friends
+  - if doing so, CPU throttling in CI will make tests flaky, [example flake](https://github.com/gardener/gardener/issues/5410)
+  - use fake clocks instead, [example PR](https://github.com/gardener/gardener/pull/4569) 
+- use the same client schemes that are also used by production code to avoid subtle bugs/regressions: [example PR](https://github.com/gardener/gardener/pull/5469/commits/b0d9e2baeaee246c6eb466a181df35adca4a6ada)
+- use the debugger to ensure your test code is actually asserting the right thing
```
Comment (ialidzhikov) [inline docs/development/testing.md]: Isn't this too abstract and kind of impossible to impose? Should we rather use more specific assertions to make sure that the corresponding branch in the code is tested? For example, prefer:
```go
	Expect(err).To(MatchError("foo"))
```

over

```go
	Expect(err).To(HaveOccurred())
```

File: docs/development/testing.md [hunk 9]
```diff
@@ -1,26 +1,194 @@
...
+    - add an interface, which your code depends on
+    - write a fake and a real implementation (similar to `clock.Clock.Sleep`)
+    - the real implementation calls the actual function (`clock.RealClock.Sleep` calls `time.Sleep`)
+    - the fake implementation does whatever you want it to do for your test (`clock.FakeClock.Sleep` waits until the test code advanced the time)
+- use constants in test code with care
+  - typically, you should not use constants from the same package as the tested code, instead use literals
+  - if the constant value is changed, tests using the constant will still pass, although the "specification" is not fulfilled anymore
+  - there are cases where it's fine to use constants, but keep this caveat in mind when doing so
+- creating sample data for tests can be a high effort
+  - if valuable, add a package for generating common sample data, e.g. Shoot/Cluster objects
```
Comment (ialidzhikov) [inline docs/development/testing.md:64]: We could also mention about the "testdata" dir - it can be useful for some cases (helm charts, yaml manifests, etc). From https://pkg.go.dev/cmd/go/internal/test:
> The go tool will ignore a directory named "testdata", making it available
to hold ancillary data needed by the tests.

Example PR - https://github.com/gardener/gardener/pull/2140

**Comment by timebertt** [comment]
Very good suggestions @ialidzhikov, thanks a lot!
I included them in the document. PTAL :)

File: docs/development/testing.md [hunk 10]
```diff
@@ -1,26 +1,206 @@
...
+- what purpose each kind of test serves
+- how to best write tests that are correct, stable, fast and maintainable
+- how to debug tests that are not working as expected
+
+The document is aimed towards developers that want to contribute code and need to write tests, as well as maintainers and reviewers that review test code.
+It serves as a common guide that we commit to follow in our project to ensure consistency in our tests, good coverage for high confidence and good maintainability.
+
+The guidelines are not meant to be absolute rules.
+Always apply common sense and adapt the guideline if it doesn't make much sense for some cases.
+If in doubt, don't hesitate to ask questions during PR review (as author but also as reviewer).
```
Comment (oliver-goetz) [inline docs/development/testing.md]: Nit:
```suggestion
If in doubt, don't hesitate to ask questions during PR review (as an author but also as a reviewer).
```

File: docs/development/testing.md [hunk 11]
```diff
@@ -1,26 +1,206 @@
...
+Add new learnings as soon as we make them!
+
+Generally speaking, **tests are a strict requirement for contributing new code**.
+If you touch code that is currently untested, you need to add tests for the new cases that you introduce as a minimum.
+Ideally though, you would add the missing test cases for the current code as well (**boy scout rule** -- "always leave the campground cleaner than you found it").
+
+## Writing Tests (Relevant for All Kinds)
+
+- we follow BDD-style testing principles and use [Ginkgo](https://onsi.github.io/ginkgo/) along with [Gomega](http://onsi.github.io/gomega/)
```
Comment (oliver-goetz) [inline docs/development/testing.md]: Nit:
```suggestion
- we follow BDD(Behavior-driven development) testing principles and use [Ginkgo](https://onsi.github.io/ginkgo/) along with [Gomega](http://onsi.github.io/gomega/)
```

File: docs/development/testing.md [hunk 12]
```diff
@@ -1,26 +1,206 @@
...
+- introduce helper functions for assertions to make test more readable where applicable: [example test](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/test/integration/gardenlet/shootsecret/controller_test.go#L323-L331)
+- introduce custom matchers to make tests more readable where applicable: [example matcher](https://github.com/gardener/gardener/blob/2eb54485231408cbdbabaa49812572a07124364f/pkg/utils/test/matchers/matchers.go#L51-L57)
+- don't rely on accurate timing of `time.Sleep` and friends
+  - if doing so, CPU throttling in CI will make tests flaky, [example flake](https://github.com/gardener/gardener/issues/5410)
+  - use fake clocks instead, [example PR](https://github.com/gardener/gardener/pull/4569) 
+- use the same client schemes that are also used by production code to avoid subtle bugs/regressions: [example PR](https://github.com/gardener/gardener/pull/5469/commits/b0d9e2baeaee246c6eb466a181df35adca4a6ada)
```
Comment (oliver-goetz) [inline docs/development/testing.md]: The example shows how the schemes could be created, but not how they are used in tests. An example for the later would help.

**Review by oliver-goetz — COMMENTED** [review]
@acumino @ary1992 found some small things in our review today

**Comment by timebertt** [comment]
@oliver-goetz @acumino @ary1992 done. PTAL :)

**Review by acumino — COMMENTED** [review]
/lgtm

**Comment by timebertt** [comment]
Any other feedback on this PR? If not, we could get it merged by the end of the day :)
